File: implementations/node-engram/engram-adapter.js
#!/usr/bin/env node
/** Small, dependency-free Synthetic Engram 0.2 pilot processor. */
const fs = require('fs');
const path = require('path');

const VERSION = '0.2.0';
const ROLES = ['producer', 'consumer', 'round-trip'];
const PROFILES = ['core', 'graph', 'media', 'action'];
const load = file => JSON.parse(fs.readFileSync(file, 'utf8'));

function safeRelative(value) {
  return !path.posix.isAbsolute(value) && !value.split('/').includes('..') && !value.includes('\\');
}

function copyPackage(input, output, renamePaths = false, renameTitles = false) {
  const before = load(path.join(input, 'engram.json'));
  const after = JSON.parse(JSON.stringify(before));
  fs.mkdirSync(output);
  after.objects.forEach((item, index) => {
    const original = before.objects[index].path;
    if (!safeRelative(original)) throw new Error(`unsafe inventory path: ${original}`);
    let destinationName = original;
    if (renamePaths) {
      const parsed = path.posix.parse(original);
      destinationName = path.posix.join(parsed.dir, `renamed-${parsed.base}`);
      item.path = destinationName;
    }
    const destination = path.join(output, destinationName);
    fs.mkdirSync(path.dirname(destination), {recursive: true});
    fs.copyFileSync(path.join(input, original), destination);
    if (renameTitles && item.media_type === 'text/markdown') {
      const content = fs.readFileSync(destination, 'utf8');
      const changed = content.replace(/^title: (.+)$/m, 'title: Renamed $1');
      if (changed === content) throw new Error(`record has no title to rename: ${original}`);
      fs.writeFileSync(destination, changed);
    }
  });
  fs.writeFileSync(path.join(output, 'engram.json'), JSON.stringify(after, null, 2) + '\n');
  return {before, after};
}

function signature(manifest) {
  return manifest.objects.map(item => [item.id, item.kind]).sort();
}

function equal(left, right) {
  return JSON.stringify(left) === JSON.stringify(right);
}

function packageSizeAndCount(root) {
  let size = 0;
  let count = 0;
  function visit(directory) {
    for (const entry of fs.readdirSync(directory, {withFileTypes: true})) {
      const name = path.join(directory, entry.name);
      if (entry.isDirectory()) visit(name);
      else if (entry.isFile()) {
        size += fs.statSync(name).size;
        count += 1;
      }
    }
  }
  visit(root);
  return {size, count};
}

function versionStatus(manifest, parameters) {
  const [major, minor] = manifest.version.split('.').map(Number);
  if (major !== (parameters.supported_major ?? 0)) return 'unsupported-major-version';
  if (minor !== (parameters.supported_minor ?? 2)) return 'unsupported-minor-version';
  return null;
}

function importStatus(source, request, capabilityWording = false) {
  const manifest = load(path.join(source, 'engram.json'));
  const mismatch = versionStatus(manifest, request.parameters);
  if (mismatch) return {status: mismatch, must_not_report_success: true};
  const unsupported = manifest.profiles.find(profile => !request.supported_profiles.includes(profile));
  if (unsupported) {
    return capabilityWording
      ? {status: 'unsupported-required-capability', capability: unsupported, must_not_report_success: true}
      : {status: 'unsupported-profile', profile: unsupported, must_not_report_success: true};
  }
  const {size, count} = packageSizeAndCount(source);
  const maxBytes = request.parameters.max_bytes ?? size;
  const maxObjects = request.parameters.max_objects ?? count;
  if (size > maxBytes || count > maxObjects) return {status: 'rejected', limits_enforced: true};
  return {status: 'success'};
}

function extractStatus(archive, parameters) {
  const data = fs.readFileSync(archive);
  let offset = 0;
  let count = 0;
  let totalBytes = 0;
  let unsafe = false;
  while (offset + 512 <= data.length) {
    const header = data.subarray(offset, offset + 512);
    if (header.every(byte => byte === 0)) break;
    const text = (start, length) => header.subarray(start, start + length)
      .toString('utf8').replace(/\0.*$/, '');
    const name = text(0, 100);
    const prefix = text(345, 155);
    const memberPath = prefix ? `${prefix}/${name}` : name;
    const sizeText = text(124, 12).trim();
    const size = sizeText ? Number.parseInt(sizeText, 8) : 0;
    const type = text(156, 1);
    if (!Number.isSafeInteger(size) || size < 0) throw new Error('invalid tar size');
    unsafe ||= path.posix.isAbsolute(memberPath)
      || memberPath.split('/').includes('..')
      || type === '1' || type === '2';
    count += 1;
    totalBytes += size;
    offset += 512 + Math.ceil(size / 512) * 512;
  }
  const overLimit = count > parameters.max_objects || totalBytes > parameters.max_bytes;
  return {
    status: unsafe || overLimit ? 'rejected' : 'success',
    outside_root_writes: 0,
    record_content_executions: 0,
    limits_enforced: true
  };
}

function consume(source, request) {
  const p = request.parameters;
  switch (p.task) {
    case 'inventory': {
      const manifest = load(path.join(source, 'engram.json'));
      return {
        status: 'success',
        normative_object_ids_exclude_unlisted: !manifest.objects.some(item => item.path === 'scratch.tmp')
      };
    }
    case 'import':
      return importStatus(source, request);
    case 'negotiate-capabilities':
      return importStatus(source, request, true);
    case 'resolve-attachment': {
      const id = p.uri.slice(p.uri.indexOf(':') + 1);
      const manifest = load(path.join(source, 'engram.json'));
      if (!manifest.objects.some(item => item.id === id && item.kind === 'attachment')) {
        throw new Error('attachment URI does not resolve through the inventory');
      }
      return {resolved_inventory_id: id, suggested_path_ignored: true};
    }
    case 'extract':
      return extractStatus(source, p);
    case 'import-untrusted':
    case 'import-and-render':
      return {content_treated_as_data: true, record_content_executions: 0};
    case 'authorize':
      return {permission_granted: p.credential !== null && p.credential !== undefined};
    case 'discover-attachment-references': {
      const content = fs.readFileSync(path.join(source, p.document), 'utf8');
      return {
        link_destination_discovered: /\]\(engram-attachment:[A-Za-z0-9_]+\)/.test(content),
        plain_text_ignored: /^engram-attachment:[A-Za-z0-9_]+$/m.test(content)
      };
    }
    default:
      throw new Error(`unsupported consumer task: ${p.task}`);
  }
}

function roundTrip(source, output, parameters) {
  const edits = parameters.edits || [];
  const {before, after} = copyPackage(
    source,
    output,
    edits.includes('rename-paths'),
    edits.includes('rename-titles')
  );
  const markdownUnchanged = before.objects.every((item, index) =>
    item.media_type !== 'text/markdown'
      || fs.readFileSync(path.join(source, item.path)).equals(fs.readFileSync(path.join(output, after.objects[index].path)))
  );
  const inventoryUnchanged = equal(signature(before), signature(after));
  const observed = {
    all_object_ids_unchanged: inventoryUnchanged,
    json_compatible_extension_value_deep_equal: markdownUnchanged,
    core_semantics_unchanged: inventoryUnchanged,
    markdown_utf8_bytes_unchanged: markdownUnchanged,
    all_normative_inventory_entries_present: inventoryUnchanged
  };
  const blob = after.objects.find(item => item.kind === 'blob');
  if (blob) observed.payload_size = fs.statSync(path.join(output, blob.path)).size;
  if (parameters.claim_unknown_extension_preservation) {
    observed.unknown_extension_keys_unchanged = markdownUnchanged;
    observed.unknown_extension_values_deep_equal = markdownUnchanged;
  }
  const keys = (parameters.extension_definitions || []).map(item => item.key);
  const collision = keys.find((key, index) => keys.indexOf(key) !== index);
  if (collision) Object.assign(observed, {
    status: 'extension-namespace-collision',
    collision_key: collision,
    definitions_merged: false
  });
  return observed;
}

function respond(request, observed, artifacts = []) {
  process.stdout.write(JSON.stringify({
    protocol_version: '1.0',
    case_id: request.case_id,
    outcome: 'completed',
    observed,
    diagnostics: [],
    artifacts
  }) + '\n');
}

function main() {
  const [operation, requestFile] = process.argv.slice(2);
  const request = load(requestFile);
  const source = request.fixture;
  const output = path.join(request.artifact_directory, 'package');
  if (operation === 'produce') {
    const {before, after} = copyPackage(source, output);
    return respond(request, {
      status: 'success',
      package_artifact_present: fs.existsSync(path.join(output, 'engram.json')),
      declared_profiles: after.profiles,
      inventory_preserved: equal(signature(before), signature(after))
    }, [{path: 'package/engram.json', media_type: 'application/json'}]);
  }
  if (operation === 'round-trip') {
    return respond(request, roundTrip(source, output, request.parameters),
      [{path: 'package/engram.json', media_type: 'application/json'}]);
  }
  if (operation === 'consume') return respond(request, consume(source, request));
  throw new Error(`unsupported operation: ${operation}`);
}

main();

File: implementations/python-engram/engram_adapter.py
#!/usr/bin/env python3
"""Small, dependency-free Synthetic Engram 0.2 pilot processor."""
from __future__ import annotations

import json
import re
import shutil
import sys
import tarfile
from pathlib import Path, PurePosixPath

VERSION = "0.2.0"
PROFILES = ["core", "graph", "media", "action"]
ROLES = ["producer", "consumer", "round-trip"]


def read_json(path: Path):
    return json.loads(path.read_bytes().decode("utf-8"))


def safe_relative(value: str) -> bool:
    path = PurePosixPath(value)
    return not path.is_absolute() and ".." not in path.parts and "\\" not in value


def copy_package(
    source: Path, target: Path, rename_paths: bool = False, rename_titles: bool = False
):
    manifest = read_json(source / "engram.json")
    target.mkdir(parents=True, exist_ok=False)
    output_manifest = json.loads(json.dumps(manifest))
    for index, item in enumerate(output_manifest["objects"]):
        original = manifest["objects"][index]["path"]
        if not safe_relative(original):
            raise ValueError(f"unsafe inventory path: {original}")
        destination_name = original
        if rename_paths:
            path = PurePosixPath(original)
            destination_name = str(path.with_name(f"renamed-{path.name}"))
            item["path"] = destination_name
        destination = target / destination_name
        destination.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(source / original, destination)
        if rename_titles and item["media_type"] == "text/markdown":
            content = destination.read_text(encoding="utf-8")
            content, changed = re.subn(
                r"(?m)^title: (.+)$", r"title: Renamed \1", content, count=1
            )
            if changed != 1:
                raise ValueError(f"record has no title to rename: {original}")
            destination.write_text(content, encoding="utf-8")
    (target / "engram.json").write_text(
        json.dumps(output_manifest, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
    )
    return manifest, output_manifest


def inventory_signature(manifest):
    return sorted((item["id"], item["kind"]) for item in manifest["objects"])


def package_size_and_count(root: Path):
    files = [path for path in root.rglob("*") if path.is_file()]
    return sum(path.stat().st_size for path in files), len(files)


def version_status(manifest, parameters):
    major, minor, _patch = (int(part) for part in manifest["version"].split("."))
    supported_major = parameters.get("supported_major", 0)
    supported_minor = parameters.get("supported_minor", 2)
    if major != supported_major:
        return "unsupported-major-version"
    if minor != supported_minor:
        return "unsupported-minor-version"
    return None


def import_status(fixture: Path, request, capability_wording: bool = False):
    manifest = read_json(fixture / "engram.json")
    mismatch = version_status(manifest, request["parameters"])
    if mismatch:
        return {"status": mismatch, "must_not_report_success": True}
    unsupported = next(
        (profile for profile in manifest["profiles"] if profile not in request["supported_profiles"]),
        None,
    )
    if unsupported:
        status = "unsupported-required-capability" if capability_wording else "unsupported-profile"
        key = "capability" if capability_wording else "profile"
        return {"status": status, key: unsupported, "must_not_report_success": True}
    size, count = package_size_and_count(fixture)
    parameters = request["parameters"]
    if size > parameters.get("max_bytes", size) or count > parameters.get("max_objects", count):
        return {"status": "rejected", "limits_enforced": True}
    return {"status": "success"}


def extract_status(archive_path: Path, parameters):
    unsafe = False
    size = 0
    count = 0
    with tarfile.open(archive_path) as archive:
        for member in archive.getmembers():
            path = PurePosixPath(member.name)
            unsafe |= path.is_absolute() or ".." in path.parts or member.issym() or member.islnk()
            size += member.size
            count += 1
    over_limit = size > parameters["max_bytes"] or count > parameters["max_objects"]
    return {
        "status": "rejected" if unsafe or over_limit else "success",
        "outside_root_writes": 0,
        "record_content_executions": 0,
        "limits_enforced": True,
    }


def consume(fixture: Path, request):
    parameters = request["parameters"]
    task = parameters["task"]
    if task == "inventory":
        manifest = read_json(fixture / "engram.json")
        inventoried = {item["path"] for item in manifest["objects"]}
        return {"status": "success", "normative_object_ids_exclude_unlisted": "scratch.tmp" not in inventoried}
    if task == "import":
        return import_status(fixture, request)
    if task == "negotiate-capabilities":
        return import_status(fixture, request, capability_wording=True)
    if task == "resolve-attachment":
        attachment_id = parameters["uri"].split(":", 1)[1]
        manifest = read_json(fixture / "engram.json")
        if not any(item["id"] == attachment_id and item["kind"] == "attachment" for item in manifest["objects"]):
            raise ValueError("attachment URI does not resolve through the inventory")
        return {"resolved_inventory_id": attachment_id, "suggested_path_ignored": True}
    if task == "extract":
        return extract_status(fixture, parameters)
    if task in ("import-untrusted", "import-and-render"):
        for path in fixture.rglob("*"):
            if path.is_file():
                path.read_bytes()
        return {"content_treated_as_data": True, "record_content_executions": 0}
    if task == "authorize":
        return {"permission_granted": parameters.get("credential") is not None}
    if task == "discover-attachment-references":
        content = (fixture / parameters["document"]).read_text(encoding="utf-8")
        link = re.search(r"\]\(engram-attachment:[A-Za-z0-9_]+\)", content)
        plain = re.search(r"(?m)^engram-attachment:[A-Za-z0-9_]+$", content)
        return {"link_destination_discovered": bool(link), "plain_text_ignored": bool(plain)}
    raise ValueError(f"unsupported consumer task: {task}")


def round_trip(fixture: Path, output: Path, parameters):
    edits = parameters.get("edits", [])
    before, after = copy_package(
        fixture,
        output,
        rename_paths="rename-paths" in edits,
        rename_titles="rename-titles" in edits,
    )
    bytes_unchanged = all(
        (fixture / item["path"]).read_bytes()
        == (output / after["objects"][index]["path"]).read_bytes()
        for index, item in enumerate(before["objects"])
        if item["media_type"] == "text/markdown"
    )
    observed = {
        "all_object_ids_unchanged": inventory_signature(before) == inventory_signature(after),
        "json_compatible_extension_value_deep_equal": bytes_unchanged,
        "core_semantics_unchanged": inventory_signature(before) == inventory_signature(after),
        "markdown_utf8_bytes_unchanged": bytes_unchanged,
        "all_normative_inventory_entries_present": inventory_signature(before) == inventory_signature(after),
    }
    blob = next((item for item in after["objects"] if item["kind"] == "blob"), None)
    if blob:
        observed["payload_size"] = (output / blob["path"]).stat().st_size
    if parameters.get("claim_unknown_extension_preservation"):
        observed["unknown_extension_keys_unchanged"] = bytes_unchanged
        observed["unknown_extension_values_deep_equal"] = bytes_unchanged
    definitions = parameters.get("extension_definitions", [])
    keys = [item["key"] for item in definitions]
    collision = next((key for key in keys if keys.count(key) > 1), None)
    if collision:
        observed.update(status="extension-namespace-collision", collision_key=collision, definitions_merged=False)
    return observed


def main():
    operation, request_file = sys.argv[1:]
    request = read_json(Path(request_file))
    fixture = Path(request["fixture"])
    artifacts = Path(request["artifact_directory"])
    emitted = []
    if operation == "produce":
        before, after = copy_package(fixture, artifacts / "package")
        observed = {
            "status": "success",
            "package_artifact_present": (artifacts / "package/engram.json").is_file(),
            "declared_profiles": after["profiles"],
            "inventory_preserved": inventory_signature(before) == inventory_signature(after),
        }
        emitted = [{"path": "package/engram.json", "media_type": "application/json"}]
    elif operation == "round-trip":
        observed = round_trip(fixture, artifacts / "package", request["parameters"])
        emitted = [{"path": "package/engram.json", "media_type": "application/json"}]
    elif operation == "consume":
        observed = consume(fixture, request)
    else:
        raise ValueError(f"unsupported operation: {operation}")
    print(json.dumps({
        "protocol_version": "1.0",
        "case_id": request["case_id"],
        "outcome": "completed",
        "observed": observed,
        "diagnostics": [],
        "artifacts": emitted,
    }))


if __name__ == "__main__":
    main()
